### PR TITLE
feat(letseed): Add LetSeed private tracker

### LINKED ISSUE #15901: [req]: LetSeed
### Have you read our Contributing Guidelines?

- [x] I have read the Contributing Guidelines

### Is there already a request for your tracker?

- [x] I have checked older issues, open and closed

### Type

Private

### Tracker details

```markdown
- Name: LetSeed
- URL: https://letseed.org/
- Description: LetSeed is a TURKISH Private Torrent Tracker for MOVIES / TV / GENERAL
- Minimum Seeding Time and/or Minimum Seeding Ratio (if Private): 1 day
```
Enigine: TS Special Edition v.8.0
Registration: open ATM

### PR DESCRIPTION
**Pull request by rpatterson**
#### Description

This is my third new indexer definition but still includes new territory for me:

- My first `download:/before:/path: takethanks.php` indexer
- The only PHP indexer I can remember that embeds the torrent DB IDs into the details
  and download URL *paths* as opposed to query parameters

So pay particular attention to those factors when reviewing.

There are also still some `# TODO: ` comments from me to address. One is specific to
this indexer and the others are general indexer definition questions that hadn't
occurred to me before.

#### Issues Fixed or Closed by this PR

* Fixes #15901

### CODE REVIEW

File: src/Jackett.Common/Definitions/letseed.yml [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+  inputs:
+    username: "{{ .Config.username }}"
+    password: "{{ .Config.password }}"
+    logout: ""
+  error:
+    - selector: '[http-equiv="refresh"][content*="/login.php"]'
+      # TODO: This tracker uses `http-equiv="refresh"` for all login form submissions,
+      # valid or invalid, and uses the `error=#` query parameter to tell the subsequent
+      # login form what error message to display. How should this definition yield a
+      # useful error message?
```
Comment (garfield69) [inline src/Jackett.Common/Definitions/letseed.yml:115]:
```suggestion
  error:
    - selector: table:has(a[href*="login.php?error=1"])
      message:
        text: "ERROR: Incorrect username!"
    - selector: table:has(a[href*="login.php?error=4"])
      message:
        text: "ERROR: Incorrect password!"
    - selector: table:has(a[href*="login.php?error="])
      message:
        text: "ERROR: Something went wrong during login!"
```
the takelogin.php page has an error code we can use, similar to futuretorrent.yml

File: src/Jackett.Common/Definitions/letseed.yml [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    # those existing definitons wrong in this aspect?  If no, why use `selector:` for
+    # web UIs that redirect to the login form?
+
+download:
+  before:
+    path: takethanks.php
+    method: post
+    inputs:
+      torrentid: >-
+        {{ re_replace .DownloadUri.AbsolutePath "/download-torrent-([0-9]+).*" "$1"}}
```
Comment (garfield69) [inline src/Jackett.Common/Definitions/letseed.yml:133]:
```suggestion
      torrentid: "{{ .DownloadUri.Query.id }}"
```
the download block has a handy variable we can use, see https://github.com/Jackett/Jackett/wiki/Definition-format#download-variables

File: src/Jackett.Common/Definitions/letseed.yml [hunk 3]
```diff
@@ -0,0 +1,215 @@
+---
+id: letseed
+name: LetSeed
+description: "LetSeed is a TURKISH Private Torrent Tracker for MOVIES / TV / GENERAL"
+# The HTML says `lang="en-US"` but the web UI defaults to Turkish, the categories are
+# always in Turkish, and there are more UI elements when using the Turkish UI:
```
Comment (garfield69) [inline src/Jackett.Common/Definitions/letseed.yml:6]:
```suggestion
```
its ok to use the language the site is aimed for, even it if differs from the default language on the page, not a problem.

**Review by garfield69 — CHANGES_REQUESTED** [review]
this is all the time I have for this, hopefully ilike2burnthing can complete the review.

File: src/Jackett.Common/Definitions/letseed.yml (hunk 2 above)
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:133]: Well, I worked off the Prowlarr docs, but they concur with the Jackett docs you linked. Are you saying that somehow the `.Query` sub-key somehow accesses path elements. Can you clarify? Where is that documented in that link? Sorry if I'm being blind.

File: src/Jackett.Common/Definitions/letseed.yml (hunk 3 above)
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:6]: Cool, but aren't code comments in part about capturing knowledge that isn't expressed in the code?

File: src/Jackett.Common/Definitions/letseed.yml [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    - {id:  3, cat: TV, desc: "Diziler"}
+    - {id:  5, cat: TV/Foreign, desc: "Yabancı Diziler"}
+    - {id:  6, cat: TV/Other, desc: "Yerli Dizi Boxset"}
+    - {id:  4, cat: TV/Other, desc: "Yerli Diziler"}
+    - {id: 44, cat: Books, desc: "Eğitim"}
+    - {id: 46, cat: Books/EBook, desc: "E-Book"}
+    - {id: 45, cat: Books/Other, desc: "Eğitim"}
+    - {id:  1, cat: Movies, desc: "Filmler"}
+    - {id: 15, cat: Movies/UHD, desc: "4K"}
+    - {id:  2, cat: Movies/Other, desc: "Animasyonlar"}
```
[the remark this answers is not in the document]
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:24]: I wrote this using a rectangle paste. Sorry for forgetting the white space cleanup.

File: src/Jackett.Common/Definitions/letseed.yml (hunk 1 above)
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml]: I'll integrate this change when I get a moment to fire up the VPN and run though all the error conditions.

File: src/Jackett.Common/Definitions/letseed.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml]: Actually, what linters/checks/tests should be run to check indexer definition changes? I'll add a commit hook for them.

[the remark this answers is not in the document]
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml]: NM, [found them](https://github.com/rpatterson/Jackett/blob/master/azure-pipelines.yml#L363-L410).

File: src/Jackett.Common/Definitions/letseed.yml (hunk 1 above)
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml]: I think I misunderstood when this indexer's web UI uses proper 3## redirects and when it's using `http-equiv="refresh"`. In the case of POSTing the login form when there's an error, it *does* do a proper `302.Redirect` so we *should* be able to use this selector. My test bed has been banned again despite rotating my VPN connection, however, so I can't confirm this right now. Someone else can test it or I'll try again soon another day

Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml]: > I think I misunderstood when this indexer's web UI uses proper 3## redirects and when it's using `http-equiv="refresh"`.

Nope, it's *both*! I just did a test in the Prowlarr edit form where I entered the wrong captcha and I got a proper `302.Redirect` response. Then I did a test where I entered the right captcha and the wrong password and I got a `http-equiv="refresh"` response. I don't know if it's stable depending on the error, but I'm not going to try and confirm that given how easy it is to hit a ban. Trying to work it out somehow.

Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml]: Thinking about this further, it would be good for a web scraping framework to treat a `http-equiv="refresh"` "redirect" as equivalent to a proper `302.Redirect` by default, probably with an option to disable. Maybe worth a feature request?

Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml]: Ok, I used both the selector for proper `302.Redirect` redirects *and* the 4 selectors for `http-equiv="refresh"` "redirects" and it worked for me when I tested:

- wrong captcha
- wrong passphrase
- wrong usename
- correct credentials and captcha

In this round of testing, *none* of the login form `POST`s got a proper `302.Redirect` response, so my best guess is that maybe the indexer's web UI switches to `http-equiv="refresh"` "redirects" after a failure. Regardless, it's definitely not stable for the error type. IOW, the same error can result in either response type.

At any rate, according to my testing this is now resolved, but it feels so fragile I'm leaving this conversation unresolved until someone else can test.

**Comment by rpatterson** [comment]
Ok, I think all the requested changes have been made and confirmed in my testing except those in unresolved conversations where I'm waiting for followup. LMK if I missed anything.

File: src/Jackett.Common/Definitions/letseed.yml (hunk 3 above)
Comment (ilike2burnthing) [reply src/Jackett.Common/Definitions/letseed.yml:6]: When useful, yes, but this comment doesn't provide a benefit here.

File: src/Jackett.Common/Definitions/letseed.yml [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+settings:
+  - name: username
+    type: text
+    label: Username
+  - name: password
+    type: password
+    label: Password
+  - name: freeleech
+    type: checkbox
+    label: Filter freeleech only
```
Comment (ilike2burnthing) [inline src/Jackett.Common/Definitions/letseed.yml:71]:
```suggestion
    label: Search freeleech only
```

File: src/Jackett.Common/Definitions/letseed.yml [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    search_type: "{{ if .Query.IMDBID }}t_genre{{ else }}t_name{{ end }}"
+    # does not support multi category searching so defaulting to all.
+    category: 0
+    # yes, no
+    include_dead_torrents: yes
+    sort: "{{ if .Config.freeleech }}free{{ else }}{{ .Config.sort }}{{ end }}"
+    order: "{{ if .Config.freeleech }}asc{{ else }}{{ .Config.type }}{{ end }}"
+
+  rows:
+    selector: "table.sortable tr:has(a[href*=\"/download-torrent-\"]){{ if .Config.freeleech }}:has(img[src$=\"/freedownload.gif\"]){{ else }}{{ end }}"
```
Comment (ilike2burnthing) [inline src/Jackett.Common/Definitions/letseed.yml:179]:
```suggestion
    selector: table.sortable tr:has(a[href*="/download-torrent-"])
```

If we're using `t_genre` to search IMDB ID, then we don't need to also filter results.

**Review by ilike2burnthing — CHANGES_REQUESTED** [review]
Just some quick points as I need to run, but will have a look later once I get my account set up and access the site.

File: src/Jackett.Common/Definitions/letseed.yml (hunk 5 above)
Comment (ilike2burnthing) [reply src/Jackett.Common/Definitions/letseed.yml:71]: Ah kk, forgot that this site engine orders results by freeleech and not, rather than only showing freeleech. Ignore this.

**Review by ilike2burnthing — APPROVED** [review]

File: src/Jackett.Common/Definitions/letseed.yml (hunk 3 above)
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:6]: On the contrary, this comment would have been very useful:

1. A contributor who doesn't read Turkish is adding a definition
2. They click the button for english in the top right on the first page they see, the login form
3. From then on, they see the english UI and never know there are more UI elements in Turkish once logged in

Exactly the kind of surprising behavior that becomes lost knowledge without a comment.

File: src/Jackett.Common/Definitions/letseed.yml [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+  error:
+    - selector: >-
+        #main table:has(td:contains("Bir hata oluştu"))
+    # Sometimes the login form POST gets a proper `302.Redirect`, but sometimes it gets
+    # a `http-equiv="refresh"` "redirect" instead:
+    - selector: >-
+        [http-equiv="refresh"][content*="/login.php?error=1"]
+      message:
+        text: >-
+          HATA: Geçersiz kullanıcı adı veya şifre! Lütfen tekrar eneyin veya
```
[the remark this answers is not in the document]
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:121]: Previous feedback indicated strings should be pulled from the UI whenever possible, which I like a lot. The most consistent behavior with that when static text must be used is to pull it from the indexer's web UI at the time of writing. Using the indexer's primary language in some cases and English in others is surprising UX, which is bad UX. In this case, if a user gets the captcha wrong on the first attempt and the password wrong on the second, they would get one error string in the indexer's language and the next in English, all in the same UI interaction.

To address the concern about developers that don't read turkish, I added comments with the Google translations.

File: src/Jackett.Common/Definitions/letseed.yml [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+  test:
+    path: browse.php
+    selector: a[href*="/logout.php"]
+    # TODO: While debugging false login success for this definition, I read in the docs
+    # that a redirect is considered as failure and that `selector:` is only required for
+    # web UIs that render the login form directly in the response to `path:`. For all
+    # the previous definitions I've worked on *and* the examples I copied for them, the
+    # web UIs responded with redirects but the definitions include `selector:`. Are
+    # those existing definitons wrong in this aspect?  If no, why use `selector:` for
+    # web UIs that redirect to the login form?
```
[the remark this answers is not in the document]
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:156]: That's not the bit I reference in my question:

> I read in the docs that a redirect is considered as failure and that `selector:` is only required for web UIs that render the login form directly in the response to `path:`.

From the same paragraph of the same docs you quoted:

> If a tracker will just show the login form (no redirect) you'll have to specify a selector too.

File: src/Jackett.Common/Definitions/letseed.yml (hunk 3 above)
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:6]: > Useful for what exactly?

Useful for the cases that code comments are useful for and thus are best practice:

- a different contributor comes along without the same domain knowledge in their head as the current maintainers
- those that had that domain knowledge have forgotten it over the years

[IOW](https://github.com/Jackett/Jackett/pull/15902#discussion_r1989897009):

> the kind of surprising behavior that becomes lost knowledge without a comment

For example, see the example from [my previous comment](https://github.com/Jackett/Jackett/pull/15902#discussion_r1989897009):

> 1. Another contributor who doesn't read Turkish is modifying this definition
> 2. They click the button for english in the top right on the first page they see, the login form
> 3. From then on, they see the english UI and never know there are more UI elements in Turkish once logged in

File: src/Jackett.Common/Definitions/letseed.yml [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+    type: image
+    selector: img#regimage
+    input: imagestring
+  inputs:
+    username: "{{ .Config.username }}"
+    password: "{{ .Config.password }}"
+    logout: ""
+  error:
+    - selector: >-
+        #main table:has(td:contains("Bir hata oluştu"))
```
[the remark this answers is not in the document]
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:112]: Enforcing specific styles is valuable when they don't get in the way of features and services provided by the languages, frameworks, tools, etc.. Once style rules exclude useful features, services, and [practices](https://github.com/Jackett/Jackett/pull/15887#discussion_r1980552811), they become wasteful, regressive dogma. For this use case, encapsulating a different syntax inside YAML, that's exactly the kind of thing [YAML's block scalars](https://yaml-multiline.info/#block-scalars) features are valuable for.

In this specific example, `selector: #main ...` doesn't work because the CSS `id="..."` selector syntax is interpreted as a YAML comment. In another example, `selector: [attr="..."]...` doesn't work because the CSS `attr="..."` attribute selector syntax is interpreted as a YAML array/list. Yes, both of these *can* be done using quotes, but then contributors have to think about which quotes to use and they have to remember to escape any quotes of that type. The behavior of both the unquoted and quoted styles are surprising and waste contributor time. The quoted style is also much less readable, contributors need to interpret two levels of quotes and escapes. None of that is necessary when we use the core features provided by the syntax we use, YAML's block scalars.

[the remark this answers is not in the document]
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:112]: > Seems like this specific contributor doesn't take into consideration that much older versions of Prowlarr can run definitions from today

You're right that I didn't take that into consideration, but only because I was unaware of that. Thanks for making me aware of that! Where are the supported versions of Prowlarr documented? I'd be happy to test.

> it won't go well with older versions of YamlDotNet.

Good thing to worry about! YAML block scalars are [a part of YAML v1.1](https://yaml.org/spec/1.1/current.html#id926597) finalized in 2005-01-18, so 20+ years ago. Looking at YamlDotNet's docs, YAML v1.1 is [the oldest version they claim conformance with](https://github.com/aaubry/YamlDotNet?tab=readme-ov-file#yamldotnet). Searching the release notes, the oldest mentions of block scalars start 2019-09-28 in v6.1.2 and all are about test cases and edge case bug fixes, not about *introducing/adding* support. In fact, the oldest PR merge listed in the release notes as regarding block scalars, specifically says [that converting block scalars to quotes strings is](https://github.com/aaubry/YamlDotNet/pull/540#issue-729672876):

>> harder to read and edit

So not only can I find no reason to expect YamlDotNet to have problems with block scalars, but I *do* find evidence of long term support and indication that the project thinks similarly about the value of block scalars. Maintaining *some* support for old software can be good, it can also definitely be bad and there's a threshold beyond which it can't be used to justify not adopting features and current best practices. That threshold varies depending on the domain, but 20+ years should be more than enough for most domains and certainly Jackett's/Prowlarr's. Put another way, I think testing support for 20+ year old YAML features is much less risky than indexer definition format changes.

> Are you going to fix those too or just generate more work for us when syncing definitions to Prowlarr?

Of course not, I'm just learning indexer definitions and I'm not proficient with the level of the stack at which such a fix would need to be implemented. Its possible to assess the project's resources to be insufficient to make a change *and* argue that it's a desirable change. That said, I would be happy to shepherd a test change through a beta release and revert if problems occur (or even if not). It would be valuable just to confirm where YAML support is fragile and gather technical details about such issues.

How *do* Jackett and Prowlarr assess the impact of changes to the indexer definition format? @B briefly mentioned such issues in Discord and I'd like to understand more completely. I'm assuming that's what the `./definitions/v##/` directories are for? From the docs it looks like [Prowlarr uses a `develop` -> `master` release cadence](https://wiki.servarr.com/prowlarr/faq#how-do-i-update-prowlarr). How does Prowlarr coordinate indexer definition changes with the introduction of support in `develop`? I'm guessing that a Prowlarr release ignores definitions from version directories newer than a latest supported definition version that is baked into the release, such that the latest `master` release would ignore the new  `./definitions/v12/*.yml` while the `develop` release will use them?

I don't find [documentation about any sort of beta release cycle for Jackett](https://github.com/Jackett/Jackett/blob/master/CONTRIBUTING.md), however, and that seems backed up by the VCS logs where I only see a `master` branch and the version tags seem to be linear and promptly follow changes. How does Jackett do beta testing of indexer definition format changes?

> don't be that new guy that changes coding styles on the first day just because that's what he's used to

Nope, not on the first day and not *just* because the change is better. I only used the block scalar here and started this discussion here because it's *relevant* to a selector that starts with `#` and *after* I'd run into other issues on *previous* PRs for selectors that start with `[`. I very much wanted to use block scalars for embedding other syntaxes in YAML, not just CSS selectors, from the first time I modified an indexer definition, but I didn't until real changes reiterated their value and in more than one case.  IOW, the change in style is motivated by real issues encountered in non-style changes, not simply style preference.

File: src/Jackett.Common/Definitions/letseed.yml (hunk 3 above)
Comment (rpatterson) [reply src/Jackett.Common/Definitions/letseed.yml:6]: > It's mostly to indicate the language of the tracker in terms what content is there, not what languages are supported.

Ah, that's a good clarification and totally makes sense, thanks! That would be good to include in the docs, that the `language:` key should reflect the language of the content and not the UI or other language variations.

That said, the value of the comment isn't to clarify for future contributors what language most of the content is in, it's to save future contributors time and confusion about how significantly the UI changes depending on the language selected.
